### PR TITLE
Parse let and const as var

### PR DESCRIPTION
**Pull request by Yuyupo**
Co-authored-by: Tamas Keri tkeri@inf.u-szeged.hu
JerryScript-DCO-1.0-Signed-off-by: Daniella Barsony bella@inf.u-szeged.hu

### CODE REVIEW

**Comment by Yuyupo** [comment]
This is not a feature just a compatibility thing, so for example one can just copy code from StackOverflow. They do work as a `var`.

**Comment by dbatyai** [comment]
As @Yuyupo said, this is a temporary "solution" for allowing let/const in the source, until we provide a proper implementation for them.

Nevertheless, as you pointed out this can create different results than what people would expect, therefore I would argue that this should not be enabled by default. It must be entirely optional, and users must be conscious about using it.

File: jerry-core/parser/js/js-parser-statm.c [hunk 1]
```diff
@@ -386,13 +394,22 @@ parser_parse_var_statement (parser_context_t *context_p) /**< context */
       parser_parse_expression (context_p,
                                PARSE_EXPR_STATEMENT | PARSE_EXPR_NO_COMMA | PARSE_EXPR_HAS_LITERAL);
     }
+#if ENABLED (JERRY_ES2015_LET_CONST_AS_VAR)
+    else
+    {
+      if (is_const)
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser-statm.c]: Please merge this if statement into an else if.

File: jerry-core/parser/js/js-parser-statm.c [hunk 2]
```diff
@@ -2300,6 +2336,8 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
         && context_p->token.type != LEXER_LEFT_BRACE
         && context_p->token.type != LEXER_RIGHT_BRACE
         && context_p->token.type != LEXER_KEYW_VAR
+        && context_p->token.type != LEXER_KEYW_LET
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser-statm.c]: Missing guards.

File: jerry-core/parser/js/js-parser-statm.c [hunk 3]
```diff
@@ -2319,6 +2357,8 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
...
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser-statm.c]: Ditto.

File: jerry-core/parser/js/js-parser-statm.c (hunk 1 above)
Comment (rerobika) [reply jerry-core/parser/js/js-parser-statm.c]: Also, this check can be incorrect inside a for in/of loop. Therefore, some kind of information need to be passed to this function ( as function argument/parser context status flag) to represent whether this function has been invoked during the for in/of statement initializer part parsing or not.

File: jerry-core/parser/js/js-parser-statm.c [hunk 4]
```diff
@@ -334,9 +334,17 @@ parser_parse_enclosed_expr (parser_context_t *context_p) /**< context */
  * Parse var statement.
  */
 static void
-parser_parse_var_statement (parser_context_t *context_p) /**< context */
+parser_parse_declaration_statement (parser_context_t *context_p) /**< context */
```
Comment (yorkie) [inline jerry-core/parser/js/js-parser-statm.c]: The original `parser_parse_var_statement` contains the initialization parsing with the keyword `var`, so the changed name seems a little bit confusion to me :)

File: jerry-core/parser/js/js-parser-statm.c [hunk 5]
```diff
@@ -1053,8 +1072,14 @@ parser_parse_for_statement_start (parser_context_t *context_p) /**< context */
     parser_save_range (context_p, &range, context_p->source_end_p);
     parser_set_range (context_p, &start_range);
     lexer_next_token (context_p);
-
-    if (context_p->token.type == LEXER_KEYW_VAR)
+#if ENABLED (JERRY_ES2015_LET_CONST_AS_VAR)
+    bool is_var_let_const = ((context_p->token.type == LEXER_KEYW_VAR)
+                               || (context_p->token.type == LEXER_KEYW_LET)
+                               || (context_p->token.type == LEXER_KEYW_CONST));
```
Comment (yorkie) [inline jerry-core/parser/js/js-parser-statm.c]: How about creating a MACRO for this kind of check?

**Comment by yorkie** [comment]
> Nevertheless, as you pointed out this can create different results than what people would expect, therefore I would argue that this should not be enabled by default. It must be entirely optional, and users must be conscious about using it.

The PR attempts to implement this feature in wrong way IMHO, optional feature doesn't mean an incompatible behavior with ECMA.

Block scoping seems better to implement in this PR, a simple way to get started is to make the `lex_env_p` of `vm_frame_ctx_t` as a stack, then at `const/let` declaration or initialization, it SET_POINTER to the top `lex_env_p`.

**Comment by dbatyai** [comment]
Unfortunately it's not that simple. If we were to implement block scoping on top of the current parser architecture, it would cause a significant increase in memory usage. We will first need to make adjustments in the parser, so that we can provide a more efficient implementation.

So, it is not the intentions of this PR to pave the way for block scoping support, instead we would like to remedy previous requests for parsing let/const, even if it's only with var semantics (#2402). As we see, this is a fairly common practice among similar engines, and since it doesn't take a lot of effort to implement, we at least wanted to have the option to allow it.

Altough I agree that non-conformant behaviour is not preferable as it can cause confusion, allowing people to optionally enable a feature like this doesn't cause any harm, as long as they are aware of the consequences. If people feel like it would cause problems in their code, they can stick to using the default configuration, and everything will work as expected.
 
Lastly, I just wanted to state that we _are_ working on implementing block-scoping in the proper way, as it is something we would like to have.

**Comment by yorkie** [comment]
> Unfortunately it's not that simple. If we were to implement block scoping on top of the current parser architecture, it would cause a significant increase in memory usage.

Hi @dbatyai could you please point out which parts will be increased significantly in memory? What I can figure it out is the lexical environments only :)

> Altough I agree that non-conformant behaviour is not preferable as it can cause confusion, allowing people to optionally enable a feature like this doesn't cause any harm, as long as they are aware of the consequences.

How about writing a document about this MACRO and its current behavior, due to a feature that's incompatible with ECMA standard would hart user in anyway.

> Lastly, I just wanted to state that we are working on implementing block-scoping in the proper way, as it is something we would like to have.

It sounds great that your team are working on this feature! BTW, the reason that I did that comment is I really think a wrong behavior is not a feature indeed, and we should not implement a "feature" that just for convenient copy & paste from StackOverflow just like @Yuyupo said.

**Pull request closed without merging**